### DISCUSSION TITLE
Multiple loggers using the same file

### DISCUSSION THREAD

**Discussion by luxapana** [gabime/spdlog#2151] [Q&A]
Is this usage valid:

```
 auto file_logger = spdlog::rotating_logger_mt("module-1", "./logs/gateway.log", 1048576 * 5, 3);
 auto file_logger2 = spdlog::rotating_logger_mt("module-2", "./logs/gateway.log", 1048576 * 5, 3);

// and in module-1
spdlog::get("module-1")->template info("started a http server");

// and in module-2
spdlog::get("module-2")->template info("started a TCP server");
```

Basically both loggers point to the same file.
I tested this and it seem to work but wanted to check if this is a valid usage before going forward.

**Comment by tt4g** [comment]
`spdlog::logger` can share sinks.

```cpp
auto rotating_file_sink = std::make_shared<spdlog::sinks::rotating_file_sink_mt>("./logs/gateway.log", 1048576 * 5, 3);

auto file_logger = std::make_shared<spdlog::logger>("module-1", rotating_file_sink);
spdlog::initialize_logger(file_logger);

auto file_logger2 = std::make_shared<spdlog::logger>("module-2", rotating_file_sink);
spdlog::initialize_logger(file_logger2);
```

See Wiki: https://github.com/gabime/spdlog/wiki/4.-Sinks#available-sinks

**Reply by congthuyen94** [reply]
how to make **rotating_file_sink**  visible between multiple file

**Reply by tt4g** [reply]
Create a different `spdlog::sinks::rotating_file_sink` for each file.

**Reply by congthuyen94** [reply]
I have many modules (multiple files). I want to  write all logs of these modules into one file sink
is it possilbe?

**Reply by tt4g** [reply]
There is no support (single sink outputs to multiple files).
But logger can have multiple sinks.

Wiki: https://github.com/gabime/spdlog/wiki/2.-Creating-loggers#creating-loggers-with-multiple-sinks

```cpp
std::vector<spdlog::sink_ptr> sinks;
sinks.push_back(std::make_shared<spdlog::sinks::rotating_file_sink_mt>("log1.txt", 1024 * 1024, 10));
sinks.push_back(std::make_shared<spdlog::sinks::rotating_file_sink_mt>("log2.txt", 1024 * 1024, 10));

auto combined_logger = std::make_shared<spdlog::logger>("name", std::begin(sinks), std::end(sinks));

//register it if you need to access it globally
spdlog::register_logger(combined_logger);
```

**Comment by gabime** [comment]
> spdlog::logger can share sinks.

Right, this is the correct way for writing to same file from multiple loggers and @manushawijekoon’s example is not valid. Also, logger->clone() can be used.

**Reply by sketch34** [reply]
Can `async_logger` share sinks?

**Reply by tt4g** [reply]
> Can `async_logger` share sinks?

@sketch34 Yes.
See Wiki: https://github.com/gabime/spdlog/wiki/2.-Creating-loggers#creating-loggers-manually
